From a6422f800ce22cf88291b7b1dc8f05513bb943eb Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Fri, 16 Oct 2009 10:26:33 +0800
Subject: [PATCH 2/3] oprofile:sparc64: fix to oprofile backtrace causing deadlock

backtrace into user space may trigger a page fault when an
oprofile event happens in kernel space. This patch guarantees
that kernel gets out of page fault exception along a fast path
instead of trying to hold mmap lock to traverse vma to verify
the address causing page fault exception. In rare situation
mmap lock has been held in advance like brk, which may lead to
a deadlock.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
Signed-off-by: Dave Lerner <dave.lerner@windriver.com>
---
 arch/sparc64/oprofile/backtrace.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/arch/sparc64/oprofile/backtrace.c b/arch/sparc64/oprofile/backtrace.c
index 2d8278d..666fdad 100644
--- a/arch/sparc64/oprofile/backtrace.c
+++ b/arch/sparc64/oprofile/backtrace.c
@@ -13,8 +13,8 @@
 #include <linux/errno.h>
 #include <linux/init.h>
 #include <linux/sched.h>
+#include <linux/uaccess.h>
 #include <asm/stacktrace.h>
-#include <asm/uaccess_64.h>
 
 #include "../kernel/kstack.h"
 
@@ -230,7 +230,7 @@ static inline void get_user_frame(struct StackState *state, unsigned long fp)
 	if (test_thread_flag(TIF_32BIT)) {
 		ufp = fp + offsetof(struct sparc_stackf32, fp);
 		addr = (void __user *)ufp;
-		ret = copy_from_user(fppc, addr, sizeof(fppc));
+		ret = probe_kernel_read(fppc, addr, sizeof(fppc));
 		/*
 		 * Even if the copy failed, it's OK to do this,
 		 * and it saves testing again later.
@@ -240,7 +240,7 @@ static inline void get_user_frame(struct StackState *state, unsigned long fp)
 	} else {
 		ufp = fp + STACK_BIAS + offsetof(struct sparc_stackf, fp);
 		addr = (void __user *)ufp;
-		ret = copy_from_user(&state->frame, addr, sizeof(struct frame));
+		ret = probe_kernel_read(&state->frame, addr, sizeof(struct frame));
 	}
 	if (unlikely(ret)) {
 #ifdef DEBUG_OPROFILE
-- 
1.6.5.rc3

